caffe2/operators/pool_op.cc

```diff
@@ -1,35 +1,603 @@
+    EigenArrayMap<T> Y_arr = kOrder == StorageOrder::NCHW
+        ? EigenArrayMap<T>(Y_ptr, Y_size, C)
+        : EigenArrayMap<T>(Y_ptr, C, Y_size);
+    for (int i = 0; i < Y_size; ++i) {
+      const int l = std::max(i * stride - pad, 0);
+      const int r = std::min(i * stride - pad + kernel, X_size);
+      ComputeMaxPool1D<T, kOrder>(C, l, r, i, X_arr, &Y_arr);
+    }
+    X_ptr += C * X_size;
+    Y_ptr += C * Y_size;
+  }
+}
+
+template <typename T, StorageOrder kOrder>
+void RunMaxPool2D(
+    const int N,
+    const int C,
+    const int X_H,
+    const int X_W,
+    const int Y_H,
+    const int Y_W,
+    const int kernel_h,
+    const int kernel_w,
+    const int stride_h,
+    const int stride_w,
+    const int pad_t,
+    const int pad_l,
+    const T* X,
+    T* Y) {
+  const int X_HxW = X_H * X_W;
+  const int Y_HxW = Y_H * Y_W;
+  EigenVectorArrayMap<T>(Y, N * C * Y_HxW)
+      .setConstant(std::numeric_limits<T>::lowest());
+  const T* X_ptr = X;
+  T* Y_ptr = Y;
+  for (int n = 0; n < N; ++n) {
+    ConstEigenArrayMap<T> X_arr = kOrder == StorageOrder::NCHW
+        ? ConstEigenArrayMap<T>(X_ptr, X_HxW, C)
+        : ConstEigenArrayMap<T>(X_ptr, C, X_HxW);
+    EigenArrayMap<T> Y_arr = kOrder == StorageOrder::NCHW
+        ? EigenArrayMap<T>(Y_ptr, Y_HxW, C)
+        : EigenArrayMap<T>(Y_ptr, C, Y_HxW);
+    for (int h = 0; h < Y_H; ++h) {
+      const int t = std::max(h * stride_h - pad_t, 0);
+      const int b = std::min(h * stride_h - pad_t + kernel_h, X_H);
+      for (int w = 0; w < Y_W; ++w) {
+        const int l = std::max(w * stride_w - pad_l, 0);
+        const int r = std::min(w * stride_w - pad_l + kernel_w, X_W);
+        const int y = h * Y_W + w;
+        ComputeMaxPool2D<T, kOrder>(C, X_W, t, b, l, r, y, X_arr, &Y_arr);
+      }
+    }
+    X_ptr += C * X_HxW;
+    Y_ptr += C * Y_HxW;
+  }
+}
+template <typename T, StorageOrder kOrder>
+void RunMaxPool3D(
+    const int N,
+    const int C,
+    const int X_D,
+    const int X_H,
+    const int X_W,
+    const int Y_D,
+    const int Y_H,
+    const int Y_W,
+    const int kernel_d,
+    const int kernel_h,
+    const int kernel_w,
+    const int stride_d,
+    const int stride_h,
+    const int stride_w,
+    const int pad_p,
+    const int pad_t,
+    const int pad_l,
+    const T* X,
+    T* Y) {
+  const int X_HxW = X_D * X_H * X_W;
+  const int Y_HxW = Y_D * Y_H * Y_W;
+  EigenVectorArrayMap<T>(Y, N * C * Y_HxW)
+      .setConstant(std::numeric_limits<T>::lowest());
+  const T* X_ptr = X;
+  T* Y_ptr = Y;
+  for (int n = 0; n < N; ++n) {
+    ConstEigenArrayMap<T> X_arr = kOrder == StorageOrder::NCHW
+        ? ConstEigenArrayMap<T>(X_ptr, X_HxW, C)
+        : ConstEigenArrayMap<T>(X_ptr, C, X_HxW);
+    EigenArrayMap<T> Y_arr = kOrder == StorageOrder::NCHW
+        ? EigenArrayMap<T>(Y_ptr, Y_HxW, C)
+        : EigenArrayMap<T>(Y_ptr, C, Y_HxW);
+    for (int d = 0; d < Y_D; ++d) {
+      const int p = std::max(d * stride_d - pad_p, 0);
```

Comment:
Nit: maybe we can have some inline function or macro to represent the following two common patterns:
- max(x * stride - pad, 0)
- min(x * stride - pad + kernel, X)